lib/BackgroundJob/MailNotificationSender.php

```diff
@@ -0,0 +1,192 @@
+
+	/** @var \OC\Group\Manager|IGroupManager  */
+	private $groupManager;
+
+	/** @var Manager  */
+	private $userManager;
+
+	/** @var NotificationManager  */
+	private $notificationManager;
+
+	/** @var IConfig  */
+	private $config;
+
+	/** @var IMailer  */
+	private $mailer;
+
+	/** @var NotificationMailerAdapter  */
+	private $mailerAdapter;
+
+	/** @var IURLGenerator  */
+	private $urlGenerator;
+
+	/** @var IRequest  */
+	private $request;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(IManager $shareManager = null,
+								IGroupManager $groupManager = null,
+								Manager $userManager = null,
+								NotificationManager $notificationManager = null,
+								NotificationMailerAdapter $mailerAdapter = null,
+								IMailer $mailer = null,
+								IConfig $config = null,
+								IURLGenerator $urlGenerator = null,
+								IRequest $request = null,
+								ILogger $logger = null) {
+		$this->shareManager = $shareManager ? $shareManager : \OC::$server->getShareManager();
+		$this->groupManager = $groupManager ? $groupManager : \OC::$server->getGroupManager();
+		$this->userManager = $userManager ? $userManager : \OC::$server->getUserManager();
+		$this->notificationManager = $notificationManager ? $notificationManager : \OC::$server->getNotificationManager();
+		$this->urlGenerator = $urlGenerator ? $urlGenerator : \OC::$server->getURLGenerator();
+		$this->request = $request ? $request : \OC::$server->getRequest();
+		$this->logger = $logger ? $logger : \OC::$server->getLogger();
+		$this->mailer = $mailer ? $mailer : \OC::$server->getMailer();
+		$this->config = $config ? $config : \OC::$server->getConfig();
+
+		$optionsStorage = new OptionsStorage($this->config);
+		$notificationMailer = new NotificationMailer($this->notificationManager, $this->mailer, $optionsStorage);
+
+		$this->mailerAdapter = $mailerAdapter ? $mailerAdapter :
+			new NotificationMailerAdapter($notificationMailer, $this->userManager, $this->logger, $this->urlGenerator);
+	}
+
+	/**
+	 * @param $notificationURL
+	 * @return string
+	 */
+	public function fixURL($notificationURL) {
+		$url = $this->request->getServerProtocol() . '://' . $this->request->getServerHost();
+		$partURL = \explode($url, $notificationURL)[1];
+		$webRoot = $this->getArgument()['webroot'];
+		if (\strpos($partURL, $webRoot) === false) {
+			$notificationURL = $url . $this->getArgument()['webroot'] . $partURL;
+		}
+		return $notificationURL;
+	}
+
+	/**
+	 * @param $shareFullId
+	 */
+	public function sendNotify($shareFullId) {
+		$notificationList = [];
+		$maxCountSendNotification = 100;
+
+		//Check if its an internal share or not
+		try {
+			$share = $this->shareManager->getShareById($shareFullId);
+		} catch (ShareNotFound $e) {
+			return null;
+		}
+
+		$users = [];
+		if ($share->getShareType() === \OCP\Share::SHARE_TYPE_GROUP) {
+			//Notify all the group members
+			$group = $this->groupManager->get($share->getSharedWith());
```

Comment:
what if the group doesn't exist? it might have been deleted after the notification was generated